src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -1275,7 +1279,150 @@ void mongoSubCacheRefresh(void)
   //
 #endif
 
-  cacheSemGive(__FUNCTION__, "subscription cache for refresh");
+  if (!semAlreadyTaken)
+  {
+    cacheSemGive(__FUNCTION__, "subscription cache for refresh");
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* CachedSubSaved - 
+*/
+typedef struct CachedSubSaved
+{
+  long long  lastNotificationTime;
+  long long  count;
+} CachedSubSaved;
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheSync - 
+*
+* 1. Save subscriptionId, lastNotificationTime, and count for all items in cache (savedSubV)
+* 2. Refresh cache (count set to 0)
+* 3. Compare lastNotificationTime in savedSubV with the new cache-contents and:
+*    3.1 Update cache-items where 'saved lastNotificationTime' > 'cached lastNotificationTime'
+*    3.2 Remember this more correct lastNotificationTime (must be flushed to mongo) -
+*        by clearing out (set to 0) those lastNotificationTimes that are newer in cache
+* 4. Update 'count' for each item in savedSubV where count != 0
+* 5. Update 'lastNotificationTime' foreach item in savedSubV where lastNotificationTime != 0
+* 6. Free the vector created in step 1 - savedSubV
+*/
+void mongoSubCacheSync(void)
+{
+  std::map<std::string, CachedSubSaved*> savedSubV;
+
+  cacheSemTake(__FUNCTION__, "Synchronizing subscription cache");
+
+
+  //
+  // 1. Save subscriptionId, lastNotificationTime, and count for all items in cache
+  //
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  while (cSubP != NULL)
+  {
+    CachedSubSaved* cssP       = new CachedSubSaved();
+
+    cssP->lastNotificationTime = cSubP->lastNotificationTime;
+    cssP->count                = cSubP->count;
+
+    savedSubV[cSubP->subscriptionId]= cssP;
+    cSubP = cSubP->next;
+  }
+
+  LM_T(LmtMongoCacheSync, ("Pushed back %d items to savedSubV", savedSubV.size()));
+
+
+  //
+  // 2. Refresh cache (count set to 0)
+  //
+  mongoSubCacheRefresh(true);
```

Comment:
Ok, now makes sense:

```
static void mongoSubCacheRefresh(std::string database)
```

(BTW, `const std::string& database` should be use there).